### PR TITLE
feat: function calling on stream

### PR DESCRIPTION
**Pull request by venjiang**
It is fully tools-supported, both in normal response and stream response.

### CODE REVIEW

**Comment by KSemenenko** [comment]
https://github.com/ollama/ollama/issues/5796 check this one

**Comment by Halfish** [comment]
Very nice feature! Can you add an reviewer for this PR?

**Comment by SinanAkkoyun** [comment]
Hi, why is this not being reviewed/merged?

**Comment by ParthSareen** [comment]
Hey @venjiang thanks for this! Going to be picking this up in the next few days - will keep you all posted.

File: server/routes.go [hunk 1]
```diff
@@ -1431,7 +1447,59 @@ func (s *Server) ChatHandler(c *gin.Context) {
...
+			slog.Info("[tools] stream response")
+			streamResponse(c, toolCallsCh)
+			return
+		} else {
+			slog.Info("[tools] no call")
+		}
+	}
+
+	slog.Info("stream response")
+	streamResponse(c, streamCh)
```
Comment (ParthSareen) [inline server/routes.go:1502]: I think we should define a better behavior here. There's a chance this gets called twice. Will look into this block when I pick this up!

**Comment by ParthSareen** [comment]
Thanks for the contribution! Just got this feature in! https://github.com/ollama/ollama/releases/tag/v0.4.6

**Pull request closed without merging**